pkg/trace/api/version.go

```diff
@@ -0,0 +1,76 @@
+package api
+
+// Version is a dumb way to version our collector handlers
+type Version string
+
+const (
+	// v01 DEPRECATED, FIXME[1.x]
+	// Traces: JSON, slice of spans
+	// Services: deprecated
+	v01 Version = "v0.1"
+
+	// v02 DEPRECATED, FIXME[1.x]
+	// Traces: JSON, slice of traces
+	// Services: deprecated
+	v02 Version = "v0.2"
+
+	// v03
+	// Traces: msgpack/JSON (Content-Type) slice of traces
+	// Services: deprecated
+	v03 Version = "v0.3"
+
+	// v04
+	// Traces: msgpack/JSON (Content-Type) slice of traces + returns service sampling ratios
+	// Services: deprecated
+	v04 Version = "v0.4"
+
+	// v05
+	//
+	// Content-Type: application/msgpack
+	// Payload: Traces with strings de-duplicated into a dictionary.
+	// Response: Service sampling rates.
+	//
+	// The payload is an array containing exactly 2 elements:
+	//
+	// 	1. An array of all unique strings present in the payload (a dictionary referred to by index).
+	// 	2. An array of traces, where each trace is an array of spans. A span is encoded as an array having
+	// 	   exactly 12 elements, representing all span properties, in this exact order:
+	//
+	// 		 0: Service   (int)
+	// 		 1: Name      (int)
+	// 		 2: Resource  (int)
+	// 		 3: TraceID   (uint64)
+	// 		 4: SpanID    (uint64)
+	// 		 5: ParentID  (uint64)
+	// 		 6: Start     (int64)
+	// 		 7: Duration  (int64)
+	// 		 8: Error     (int32)
+	// 		 9: Meta      (map[int]int)
+	// 		10: Metrics   (map[int]float64)
+	// 		11: Type      (int)
+	//
+	// 	Considerations:
+	//
+	// 	- The "int" typed values represent the index at which the corresponding string is found in the dictionary.
+	// 	  If any of the values are the empty string, then the empty string must be added into the dictionary.
+	//
+	// 	- None of the elements can be nil. If any of them are unset, they should be given their "zero-value". Here
+	// 	  is an example of a span with all unset values:
+	//
+	// 		 0: 0                 // Service is "" (index 0 in dictionary)
```

Comment:
Should the specification be that the index `0` is always the string `""`? I see this as having the opportunity to:

- May allow some minor code simplifications or even micro speed gain (`""` can be a common string , `if 0` no need to call dictionaryString).
- Is a "sane" default which could reduce the impact of a bug. Example: if a bug is to incorrectly set an attribute to its ID, it'd then default to 0. Translating this to the string empty is safer than picking whatever got encoded first.

Reply:
Right, my arguments weren't properly weighted nor ordered. IMHO the perf argument is anecdotal. 

But I still see it as a safer specification, mostly in the case of bugs.
If a tracer has a bug which result in incorrectly not setting a value, it'd be safer and much easier to debug if it were always translated into the empty string. 

(On the more conceptual aspect, it "looks nicer" to align the neutral element when transforming one group to the other. But this is purely a personal feeling/preference.)